asn1/Stream.asn

```diff
@@ -174,4 +194,12 @@ StreamDataBlocked ::= SEQUENCE {
   maxOffset VarUInt
 }
 
+StreamReceipt ::= SEQUENCE {
+  -- Identifier of the stream
+  streamId VarUInt,
+
+  -- Receipt for verifying the total amount received on this stream
+  receipt Receipt
```

Comment:
Given that the STREAM [spec](https://interledger.org/rfcs/0029-stream/#53-frames) and [implementation](https://github.com/interledgerjs/ilp-protocol-stream/blob/master/src/packet.ts#L491-L492) indicate that
>Each frame is encoded with its 1-byte type and length prefix.

the fact that the STREAM [asn.1 definition](https://interledger.org/rfcs/asn1/Stream.asn) doesn't specify the length prefix for the various frame `SEQUENCE`s, _seems_ to suggest that `SEQUENCE`s are length prefixed.
But I'm not finding an indication either way in the asn.1 docs.

Reply:
I don't think the SEQUENCEs we use are length-prefixed because they're not `extensible`.

> A non-extensible sequence type is encoded as a presence bitmap followed by the concatenation of the encodings of the fields of the sequence type that are present in the value, in specification order. The presence bitmap is encoded as a bit string with a fixed size constraint (see 2.7, case 1), and has one bit for each field of the sequence type that has the keyword OPTIONAL or DEFAULT, in specification order.

(see [Overview_of_OER.pdf, Section 2.8](https://www.oss.com/asn1/resources/books-whitepapers-pubs/Overview_of_OER.pdf).

Though I'm curious: do we think the `Receipt` is going to change in the future and become extensible?

Reply:
Thanks for the link.
After discussing it with Ben, we'd like to follow the model of [ILP packets](https://github.com/interledger/rfcs/blob/master/asn1/InterledgerPacket.asn#L54-L59) by depending on the version byte to know what to expect for the remaining bytes (instead of making `Receipt` extensible and introducing the presence / extension presence bitmaps).
The entire receipt is length prefixed in the `StreamReceipt` frame in order for stream senders to be able to handle unsupported receipt versions.
https://github.com/interledger/rfcs/pull/570/commits/09172b7f714005bc4df682637829bf001b53a9bf

Reply:
> we'd like to follow the model of ILP packets by depending on the version byte to know what to expect for the remaining bytes

No objection here - I like this idea.

That said (and I'm no ASN expert, so there's that) -- the [current ASN](https://github.com/interledger/rfcs/commit/09172b7f714005bc4df682637829bf001b53a9bf#diff-b95395fdef50093ed11e454c0d9b050bR204) doesn't seem to have a strong connection between an actual `Receipt` and just random bytes because the type defined in the `StreamReceipt Frame` for the actual Receipt is just `VarBytes` (which is really just [OCTET STRING](https://github.com/interledger/rfcs/blob/master/asn1/GenericTypes.asn#L65)).

What if instead we more closely followed the concept of the InterledgerPacket here so it's obvious what's allowed in the `StreamReceipt` frame? 

Currently, InterlederPacket supports a bunch of different "types". In our case, we could mirror this style and just define new "types" in the future if we ever need a new Receipt version. Primarily, this would create a stronger connection between the `Receipt` field in the frame and the data allowed in it. ASN could look like this:

https://gist.github.com/sappenin/b5925e86a2cc6e8042650e6802800a7a

Reply:
Thanks for the gist.
I agree it's worth trying to associate `Receipt` with the `StreamReceiptFrame`.
A difference here from ILP and STREAM packets is that the receipt type/version is needed by the verifier but not the STREAM sender (who parses the frame). So I think including the type in the frame (as opposed to in the receipt) just adds a useless field for STREAM implementations to parse.
What do you think about this alternative which uses a CHOICE:
https://gist.github.com/wilsonianb/75ddc1b2a3780218016f1da0f28f9101

Reply:
I suppose I need to know more about what we think a future Receipt might look like. I see the comments above around a future `Receipt` having potentially a different length, but I don't quite have the context to know what that would mean. Can you elaborate there?

The reason I bring that up is that the `CHOICE` option seems unclear/complicated in terms of how to encode it (e.g., there's the tag itself, and then there appears to be different types of tags [in section 2.12](https://www.oss.com/asn1/resources/books-whitepapers-pubs/Overview_of_OER.pdf)).

Also, using `CHOICE` seems to be more geared towards payloads that should have one thing or another (e.g., `house` or `car`) -- it doesn't feel quite right for versioning (seems like extensibility is more correct there).

Wondering if anyone with more ASN.1 OER expertise should chime in here...  @adrianhopebailie  ?

> A difference here from ILP and STREAM packets is that the receipt type/version is needed by the verifier but not the STREAM sender (who parses the frame). 

Can you clarify this? In my mind the verifier could parse per-usual, but the STREAM sender wouldn't actually parse the `Receipt` if they don't need it. They can just treat this as a blob of data, no?

Reply:
I've updated the [gist](https://gist.github.com/wilsonianb/75ddc1b2a3780218016f1da0f28f9101) with a hypothetical receipts v2, but I hadn't looked at how `CHOICE` is handled in OER and agree that a tag is not desirable.

I don't think we have anything concrete in mind for what a future version might look like.
I'm mainly trying to support the possibility of future version(s) such that:
- the sender can parse a receipt blob of any length out of the `StreamReceipt` frame (without parsing the receipt itself, not even the version/type)
- the verifier can check a receipt's version/type to known if they support it.

>In my mind the verifier could parse per-usual, but the STREAM sender wouldn't actually parse the Receipt if they don't need it. They can just treat this as a blob of data, no?

That's all correct, but (using your [example gist](https://gist.github.com/sappenin/b5925e86a2cc6e8042650e6802800a7a)) the sender would need to parse the `ReceiptPacket`, where the type/version lives, in order to get at the `Receipt`.

How about this minor variation on your example
https://gist.github.com/wilsonianb/b164080a6c1b62e0e2be11d41efb8b13

The `Receipt` (formerly `ReceiptPacket`) is what's passed from the sender to the verifier.
The data in `Receipt` wouldn't be length prefixed <-- would that need to be represented differently in ASN.1?
The version number wouldn't be part of the hmac message, which should be fine.

Reply:
>The version number wouldn't be part of the hmac message, which should be fine.

Ben pointed out that could not be fine. A hypothetical subsequent version could swap the nonce and amount fields, and the receipt created as one version _could_ be submitted as the other version.

I've updated the hmac description in my gist:
https://gist.github.com/wilsonianb/b164080a6c1b62e0e2be11d41efb8b13
Does that seem workable, especially the non-length prefixed `Receipt` data field?

Reply:
@wilsonianb what's in your latest gist in the [comment](https://github.com/interledger/rfcs/pull/570#discussion_r409861591) directly above LGTM. 

> Ben pointed out that could not be fine. 

Great catch @sharafian !

Reply:
updated in https://github.com/interledger/rfcs/pull/570/commits/43039f71db1b815d54e42494aeea282ed18f5c4e